src/idpi/grib_decoder.py

```diff
@@ -91,227 +94,303 @@ def _extract_pv(pv):
+        datafiles: list[Path],
+        ref_param: str = "HHL",
+        ifs: bool = False,
+        delay: bool = False,
+    ):
+        """Initialize a grib reader from a list of grib files.
+
+        Parameters
+        ----------
+        datafiles : list[Path]
+            List of grib input filenames
+        ref_param : str
+            name of parameter used to construct a reference grid
+        ifs : bool
+            True for setting up a grib reader for IFS data
+        delay : bool
+            if True, it will (dask) delay the functions that load parameters
+
+        Raises
+        ------
+        ValueError
+            if the grid can not be constructed from the ref_param
+
+        """
+        self._datafiles = [str(p) for p in datafiles]
+        self._ifs = ifs
+        self._delayed = dask.delayed if delay else (lambda x: x)
+        if not self._ifs:
+            global _ifs_allowed
+            _ifs_allowed = False  # due to incompatible data in cache
+
+            with cosmo_grib_defs():
+                self._grid = self.load_grid_reference(ref_param)
+        else:
+            self._grid = self.load_grid_reference(ref_param)
+
+    def load_grid_reference(self, ref_param: str) -> Grid:
+        """Construct a grid from a reference parameter.
+
+        Parameters
+        ----------
+        ref_param : str
+            name of parameter used to construct a reference grid.
+
+        Raises
+        ------
+        ValueError
+            if ref_param is not found in the input dataset
+
+        Returns
+        -------
+        Grid
+            reference grid
+
+        """
+        if self._ifs:
+            mapping_path = files("idpi.data").joinpath("field_mappings.yml")
+            mapping = yaml.safe_load(mapping_path.open())
+            ref_param = mapping[ref_param]["ifs"]["name"]
+
+        fs = earthkit.data.from_source("file", [str(p) for p in self._datafiles])
+        it = iter(fs.sel(param=ref_param))
+        field = next(it, None)
+        if field is None:
+            msg = f"reference field, {ref_param=} not found in {self._datafiles=}"
+            raise RuntimeError(msg)
+        lonlat_dict = {
+            geo_dim: xr.DataArray(dims=("y", "x"), data=values)
+            for geo_dim, values in field.to_latlon().items()
+        }
+
+        grid = Grid(
+            lonlat_dict["lon"],
+            lonlat_dict["lat"],
+            *field.metadata(
+                "longitudeOfFirstGridPointInDegrees",
+                "latitudeOfFirstGridPointInDegrees",
+            ),
+        )
+
+        return grid
+
+    def _load_pv(self, pv_param: str):
+        if not self._ifs:
+            raise ValueError("load_pv only available for IFS data")
+        fs = earthkit.data.from_source("file", [str(p) for p in self._datafiles]).sel(
+            param=pv_param
+        )
+
+        for field in fs:
+            return field.metadata("pv")
+
+    def _construct_metadata(self, field: typing.Any):
```

Comment:
I tried here to set the type to earthkit.data.xxx.GribField, however that requires loading the corresponding module of earthkit which causes that our current mechanism of setting the eccodes definitions path to fail. One can see that in the previous builds of the jenkins plans for this PR.